src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs

```diff
@@ -202,13 +202,31 @@ internal Task ProduceTagsAsync_ForTestingPurposesOnly(TaggerContext<TTag> contex
 
         /// <summary>
         /// Produce tags for the given context.
+        /// Keep in sync with <see cref="ProduceTagsSynchronously(TaggerContext{TTag})"/>
         /// </summary>
         protected virtual async Task ProduceTagsAsync(TaggerContext<TTag> context)
         {
             foreach (var spanToTag in context.SpansToTag)
             {
                 context.CancellationToken.ThrowIfCancellationRequested();
-                await ProduceTagsAsync(context, spanToTag, GetCaretPosition(context.CaretPosition, spanToTag.SnapshotSpan)).ConfigureAwait(false);
+                await ProduceTagsAsync(
+                    context, spanToTag,
+                    GetCaretPosition(context.CaretPosition, spanToTag.SnapshotSpan)).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Produce tags for the given context.
+        /// Keep in sync with <see cref="ProduceTagsAsync(TaggerContext{TTag})"/>
+        /// </summary>
+        protected void ProduceTagsSynchronously(TaggerContext<TTag> context)
+        {
+            foreach (var spanToTag in context.SpansToTag)
+            {
+                context.CancellationToken.ThrowIfCancellationRequested();
+                ProduceTagsSynchronously(
```

Comment:
Just observing that the pattern here is different than the pattern in the async case. The async one has both ProduceTagsAsync* calls being virtual, whereas your synchronous case has just one. Frankly, the fact that both are virtual in the async case seems pretty fishy...any idea why that was done?